### PR TITLE
Spalart allmaras merge patch

### PR DESCRIPTION
**Pull request by roystgnr**
This replaces #299, with the "intermediate commits fail make check" problems fixed.

### CODE REVIEW

**Comment by roystgnr** [comment]
No, wait, forgot one reordering.

**Comment by roystgnr** [comment]
Okay.  d672093 will still fail, but IMHO that's a lesser sin than melding the "bugfix" and "regression gold standard update to match bugfix" commits.

File: examples/sa_airfoil/sa_airfoil.in [hunk 1]
```diff
@@ -0,0 +1,127 @@
+# Mesh related options
+[Mesh]
+   [./Read]
+      filename = 'naca_0012_medium_boundary_id.xda'
+[]
+
+[restart-options]
+   # Use the coarser mesh to give us a good initial guess for the solution
+   restart_file = '/workspace/vikram/grins_libmesh_master_opt/examples/sa_airfoil/AoA_0/turbulent_airfoil_Re_6000000.xda'
```
Comment (pbauman) [inline examples/sa_airfoil/sa_airfoil.in]: Sorry, can we make this a local path?

**Comment by pbauman** [comment]
OK, I'll pull and test on my laptop right now.

**Comment by roystgnr** [comment]
Passes "make check" for me (with gold solution generated by Vikram, so I assume it's passing for him).  I'll check on that example.

**Comment by pbauman** [comment]
I hit the following during `make check` of the `sa_2d_turbulent_channel_regression.sh` test. Forward solve completed fine, but then trying to read the xdr file:

<pre>
Assertion `n_assigned_vals == vector_length' failed.
n_assigned_vals = 3171
vector_length = 12684
</pre>

<pre>
[02:19:02][pbauman@farnsworth:~/work/research/grins/build/pr300/devel]$ module list

Currently Loaded Modules:
  1) gcc/4.9.3   2) openblas/0.2.14   3) mpich/3.1.4   4) petsc/3.5.4-opt   5) tbb/4.3   6) hdf5/1.8.15-patch1   7) boost/1.57.0   8) vtk/6.2.0   9) libmesh/master  10) cantera/2.1.2
</pre>

libMesh is not tip of master, but is not too old: e8fc261438c8a1545e4d34f242cc16a748209783
(We should have `libmesh-config --version` dump the hash.)

I'm copying the entire test output below (modulo perf log).

<pre>
[02:22:45][pbauman@farnsworth:~/work/research/grins/build/pr300/devel/test]$ ./sa_2d_turbulent_channel_regression.sh 
*** Warning, AutoPtr is deprecated and will be removed in a future library version! ./include/libmesh/auto_ptr.h, line 270, compiled Jul 24 2015 at 11:53:56 ***
 EquationSystems
  n_systems()=1
   System #0, "flow"
    Type "Implicit"
    Variables={ "u" "nutil" } 
    Finite Element Types="LAGRANGE", "JACOBI_20_00" 
    Infinite Element Mapping="CARTESIAN" 
    Approximation Orders="FIRST", "THIRD" 
    n_dofs()=662
    n_local_dofs()=662
    n_constrained_dofs()=0
    n_local_constrained_dofs()=0
    n_vectors()=2
    n_matrices()=1
    DofMap Sparsity
      Average  On-Processor Bandwidth <= 5.98792
      Average Off-Processor Bandwidth <= 0
      Maximum  On-Processor Bandwidth <= 6
      Maximum Off-Processor Bandwidth <= 0
    DofMap Constraints
      Number of DoF Constraints = 0
      Number of Node Constraints = 0

*** Warning, This code is deprecated, and likely to be removed in future library versions! ../../../../pbauman/src/solver/src/simulation.C, line 69, compiled Sep  2 2015 at 14:08:04 ***
No of walls: 2
Boundary Id: 0
Boundary Id: 2
No of walls: 2
Boundary Id: 0
Boundary Id: 2
==========================================================
List of Enabled Physics:
IncompressibleNavierStokes
IncompressibleNavierStokesSPGSMStabilization
SpalartAllmaras
SpalartAllmarasSPGSMStabilization
==========================================================
Distance Function computed.
 Mesh Information:
  elem_dimensions()={2}
  spatial_dimension()=3
  n_nodes()=3171
    n_local_nodes()=3171
  n_elem()=3000
    n_local_elem()=3000
    n_active_elem()=3000
  n_subdomains()=1
  n_partitions()=1
  n_processors()=1
  n_threads()=1
  processor_id()=0

 EquationSystems
  n_systems()=2
   System #0, "GRINS"
    Type "Implicit"
    Variables={ "u" "v" "p" "nu" } 
    Finite Element Types="LAGRANGE", "JACOBI_20_00" 
    Infinite Element Mapping="CARTESIAN" 
    Approximation Orders="FIRST", "THIRD" 
    n_dofs()=12684
    n_local_dofs()=12684
    n_constrained_dofs()=424
    n_local_constrained_dofs()=424
    n_vectors()=1
    n_matrices()=1
    DofMap Sparsity
      Average  On-Processor Bandwidth <= 34.7032
      Average Off-Processor Bandwidth <= 0
      Maximum  On-Processor Bandwidth <= 36
      Maximum Off-Processor Bandwidth <= 0
    DofMap Constraints
      Number of DoF Constraints = 424
      Number of Heterogenous Constraints= 298
      Average DoF Constraint Length= 0
      Number of Node Constraints = 0
   System #1, "distance_function"
    Type "Basic"
    Variables="distance" 
    Finite Element Types="LAGRANGE", "JACOBI_20_00" 
    Infinite Element Mapping="CARTESIAN" 
    Approximation Orders="FIRST", "THIRD" 
    n_dofs()=3171
    n_local_dofs()=3171
    n_constrained_dofs()=0
    n_local_constrained_dofs()=0
    n_vectors()=0
    n_matrices()=0
    DofMap Sparsity
      Average  On-Processor Bandwidth <= 0
      Average Off-Processor Bandwidth <= 0
      Maximum  On-Processor Bandwidth <= 0
      Maximum Off-Processor Bandwidth <= 0
    DofMap Constraints
      Number of DoF Constraints = 0
      Number of Node Constraints = 0

Assembling the System
*** Warning, This code is deprecated, and likely to be removed in future library versions! /Users/pbauman/work/software/libs/libmesh/master/gcc/4.9.3/mpich/3.1.4/petsc/3.5.4/openblas/0.2.14/tbb/4.3/boost/1.57.0/hdf5/1.8.15-patch1/vtk/6.2.0/include/libmesh/libmesh_common.h, line 494, compiled Sep  2 2015 at 14:04:14 ***
Nonlinear Residual: 0.0230132
Linear solve starting, tolerance 0.001
Linear solve finished, step 48, residual 0.0102452
Trying full Newton step
  Current Residual: 0.00828398
  Nonlinear step: |du|/|u| = 0.386619, |du| = 18.1644
Assembling the System
Nonlinear Residual: 0.00828398
Linear solve starting, tolerance 0.001
Linear solve finished, step 32, residual 0.00534235
Trying full Newton step
  Current Residual: 0.00300857
  Nonlinear step: |du|/|u| = 0.117732, |du| = 5.58259
Assembling the System
Nonlinear Residual: 0.00300857
Linear solve starting, tolerance 0.001
Linear solve finished, step 28, residual 0.00141176
Trying full Newton step
  Current Residual: 0.00111076
  Nonlinear step: |du|/|u| = 0.0410494, |du| = 1.95226
Assembling the System
Nonlinear Residual: 0.00111076
Linear solve starting, tolerance 0.001
Linear solve finished, step 23, residual 0.000250485
Trying full Newton step
  Current Residual: 0.000153599
  Nonlinear step: |du|/|u| = 0.00701056, |du| = 0.333251
Assembling the System
Nonlinear Residual: 0.000153599
Linear solve starting, tolerance 0.001
Linear solve finished, step 23, residual 7.35757e-05
Trying full Newton step
  Current Residual: 8.96147e-06
  Nonlinear step: |du|/|u| = 0.00225366, |du| = 0.107099
Assembling the System
Nonlinear Residual: 8.96147e-06
Linear solve starting, tolerance 0.001
Linear solve finished, step 23, residual 1.87772e-05
Trying full Newton step
  Current Residual: 5.08873e-07
  Nonlinear step: |du|/|u| = 0.000526548, |du| = 0.0250212
Assembling the System
Nonlinear Residual: 5.08873e-07
Linear solve starting, tolerance 0.001
Linear solve finished, step 25, residual 6.91166e-07
Trying full Newton step
  Current Residual: 2.63972e-08
  Nonlinear step: |du|/|u| = 1.71873e-05, |du| = 0.000816726
Assembling the System
Nonlinear Residual: 2.63972e-08
Linear solve starting, tolerance 0.001
Linear solve finished, step 25, residual 3.3259e-08
Trying full Newton step
  Current Residual: 1.01649e-09
  Nonlinear step: |du|/|u| = 9.24517e-07, |du| = 4.39324e-05
Assembling the System
Nonlinear Residual: 1.01649e-09
Linear solve starting, tolerance 0.001
Linear solve finished, step 24, residual 1.16321e-09
Trying full Newton step
  Current Residual: 2.7254e-11
  Nonlinear step: |du|/|u| = 3.62404e-08, |du| = 1.72212e-06
Assembling the System
Nonlinear Residual: 2.7254e-11
Linear solve starting, tolerance 0.001
Linear solve finished, step 24, residual 1.64736e-11
Trying full Newton step
  Current Residual: 3.41402e-13
  Nonlinear step: |du|/|u| = 3.88784e-10, |du| = 1.84748e-08
Assembling the System
Nonlinear Residual: 3.41402e-13
Linear solve starting, tolerance 0.001
Linear solve finished, step 25, residual 1.24546e-13
Trying full Newton step
  Current Residual: 3.15713e-15
  Nonlinear solver relative residual 1.37188e-13 > 1e-15
  Nonlinear solver converged, step 10, relative step size 3.90413e-12 < 1e-10
Assertion `n_assigned_vals == vector_length' failed.
n_assigned_vals = 3171
vector_length = 12684


Stack frames: 2
0: 0   libmesh_devel.0.dylib               0x000000011035ca80 libMesh::print_trace(std::ostream&) + 1600
1: 1   libmesh_devel.0.dylib               0x000000011035a56a libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 138
[0] src/systems/system_io.C, line 1264, compiled Jul 24 2015 at 11:55:50

*********************************************************************
READING THE FILE "../../../../pbauman/test/test_data/sa_2d_turbulent_channel_regression.xdr" FAILED.
It is possible this file contains infinite element information,
but the version string does not contain " with infinite elements"
Let's try this again, but looking for infinite element information...
*********************************************************************

ERROR: incompatible variable  has already been added for this system!
Stack frames: 2
0: 0   libmesh_devel.0.dylib               0x000000011035ca80 libMesh::print_trace(std::ostream&) + 1600
1: 1   libmesh_devel.0.dylib               0x000000011035a56a libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 138
[0] src/systems/system.C, line 1115, compiled Jul 24 2015 at 11:55:50

*********************************************************************
Well, at least we tried!
Good Luck!!
*********************************************************************
</pre>

**Comment by pbauman** [comment]
`git log` says I have 1d688e1... not sure what the deal is.

**Comment by roystgnr** [comment]
You're using pre- https://github.com/libMesh/libmesh/pull/618 libMesh.  I'm guessing that Vikram generated a regression test file without the option to output "additional" data, but the regression test doesn't use that option when reading?

**Comment by roystgnr** [comment]
The regression test does.  Let me fix that and you can try again.

**Comment by pbauman** [comment]
Ha! Apparently "libMesh not too old" was wrong. Thanks for the quick fix - testing now.

**Comment by pbauman** [comment]
Hmmm, I'm still getting the same problem. I think it has to do with reading the reference solution [here](https://github.com/roystgnr/grins/blob/spalart_allmaras_merge_patch/test/test_turbulent_channel.C#L282): I put a breakpoint right before there and could run up until that point and then next line gave the error. Going to experiment more.

**Comment by pbauman** [comment]
OK, yeah, here we go. Patch below gets the test to pass for me.

<pre>
diff --git a/test/test_turbulent_channel.C b/test/test_turbulent_channel.C
index adb5c15..864bfe6 100644
--- a/test/test_turbulent_channel.C
+++ b/test/test_turbulent_channel.C
@@ -279,7 +279,10 @@ int main(int argc, char* argv[])
 
   // Filename of file where comparison solution is stashed
   std::string solution_file = command_line("soln-data", "DIE!");
-  es_ref.read( solution_file );
+  es_ref.read( solution_file, libMesh::XdrMODE::DECODE,
+                              libMesh::EquationSystems::READ_HEADER |
+                              libMesh::EquationSystems::READ_DATA |
+                              libMesh::EquationSystems::READ_ADDITIONAL_DATA );
 
   exact_sol.attach_reference_solution( &es_ref );
</pre>

I'm guessing the "gold data" was just 'xdr' output via GRINS, which has the `WRITE_ADDITIONAL_DATA` in it. Post libMesh/libmesh#618 wouldn't require this, I'm guessing, since you guys didn't hit it? That ticket talked about the trying to read additional data that wasn't there, but this was a case of not trying to read additional data that was there.

**Comment by roystgnr** [comment]
Thanks!

**Comment by pbauman** [comment]
Thanks!

**Pull request merged**